platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/SyncPreConsensusEventWriter.java

```diff
@@ -0,0 +1,392 @@
+     * @param fileManager
+     * 		manages all preconsensus event stream files currently on disk
+     */
+    public SyncPreConsensusEventWriter(
+            final PreConsensusEventStreamConfig config, final PreConsensusEventFileManager fileManager) {
+
+        preferredFileSizeMegabytes = config.preferredFileSizeMegabytes();
+
+        averageGenerationalSpanUtilization =
+                new LongRunningAverage(config.generationalUtilizationSpanRunningAverageLength());
+        bootstrapGenerationalSpan = config.bootstrapGenerationalSpan();
+        generationalSpanOverlapFactor = config.generationalSpanOverlapFactor();
+        minimumGenerationalCapacity = config.minimumGenerationalCapacity();
+
+        this.fileManager = fileManager;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public synchronized void writeEvent(final EventImpl event) {
+        validateSequenceNumber(event);
+        if (event.getGeneration() >= minimumGenerationNonAncient) {
+            writeEventToStream(event);
+            flushIfNeeded();
+        } else {
+            event.setStreamSequenceNumber(EventImpl.STALE_EVENT_STREAM_SEQUENCE_NUMBER);
+        }
+    }
+
+    /**
+     * Make sure that the event has a valid stream sequence number.
+     */
+    private static void validateSequenceNumber(final EventImpl event) {
+        if (event.getStreamSequenceNumber() == EventImpl.NO_STREAM_SEQUENCE_NUMBER
+                || event.getStreamSequenceNumber() == EventImpl.STALE_EVENT_STREAM_SEQUENCE_NUMBER) {
+            throw new IllegalStateException("Event must have a valid stream sequence number");
+        }
+    }
+
+    /**
+     * Flush the stream if needed. Should be called after each event is written.
+     */
+    private void flushIfNeeded() {
+        // Remove all events that have already been flushed. This scenario is relevant after a file is closed.
+        while (!flushableEvents.isEmpty() && flushableEvents.peek() < lastWrittenEvent) {
+            flushableEvents.remove();
+        }
+        if (!flushableEvents.isEmpty()) {
+            final long nextFlushableEvent = flushableEvents.peek();
+            if (nextFlushableEvent == lastWrittenEvent) {
+                try {
+                    currentMutableFile.flush();
+                } catch (final IOException e) {
+                    throw new UncheckedIOException(e);
+                }
+                markEventsAsFlushed();
+                flushableEvents.remove();
+            }
+        }
+    }
+
+    /**
+     * Write an event to the file stream.
+     * @param event the event to write
+     */
+    private void writeEventToStream(final EventImpl event) {
+        try {
+            prepareOutputStream(event);
+            currentMutableFile.writeEvent(event);
+            lastWrittenEvent = event.getStreamSequenceNumber();
+        } catch (final IOException e) {
+            throw new UncheckedIOException(e);
```

Comment:
I ask myself if an illegalArgException would be better.

Reply:
We could have an IO exception if there was a file system error, completely unrelated to the argument passed in.